### ISSUE TITLE
Complex query criteria ($size + $all) doesn't work

### ISSUE THREAD

**Issue by piotrrzacki** [bwaldvogel/mongo-java-server#96]
Dear all.
I tried to integrate mongo-java-server:1.20.0 with Spring boot starter data mongodb, and this lib is working for simple criteria queries. However complex ones are not working. 

Am I doing something wrong, or this feature were not yet implemented?

Simple project to reproduce this issue:

**build.gradle**:
```groovy
plugins {
    id 'org.springframework.boot' version '2.1.8.RELEASE'
    id 'io.spring.dependency-management' version '1.0.8.RELEASE'
    id 'java'
}

group = 'example'
version = '0.0.1-SNAPSHOT'
sourceCompatibility = '1.8'

repositories {
    mavenCentral()
}

ext {
    set('springBootVersion', "2.1.8.RELEASE")
}

dependencies {
    implementation 'de.bwaldvogel:mongo-java-server:1.20.0'
    implementation "org.springframework.boot:spring-boot-starter-data-mongodb"
    implementation "org.springframework.boot:spring-boot-starter-test"
}

dependencyManagement {
    imports {
        mavenBom "org.springframework.boot:spring-boot-starter-parent:${springBootVersion}"
    }
}
```

**example.DemoApplication.java**:
```java
@SpringBootApplication
public class DemoApplication {

    public static void main(String[] args) {
        SpringApplication.run(DemoApplication.class, args);
    }

    @Bean
    public MongoTemplate mongoTemplate(MongoClient mongoClient) {
        return new MongoTemplate(mongoDbFactory(mongoClient));
    }

    @Bean
    public MongoDbFactory mongoDbFactory(MongoClient mongoClient) {
        return new SimpleMongoDbFactory(mongoClient, "test");
    }

    @Bean(destroyMethod="shutdown")
    public MongoServer mongoServer() {
        MongoServer mongoServer = new MongoServer(new MemoryBackend());
        mongoServer.bind();
        return mongoServer;
    }

    @Bean(destroyMethod="close")
    public MongoClient mongoClient(MongoServer mongoServer) {
        return new MongoClient(new ServerAddress(mongoServer.getLocalAddress()));
    }

}
```

**example.DemoApplicationTests.java**:
```java
@RunWith(SpringRunner.class)
@SpringBootTest
public class DemoApplicationTests {

    @Autowired
    MongoTemplate mongoTemplate;

    @Before
    public void before() {
        mongoTemplate.remove(new BasicQuery("{}"), TestDocument.class);
    }

    @Test
    public void complexCriteriaAreWorking() {
        // insert document with field as list
        TestDocument doc = new TestDocument();
        doc.id = UUID.randomUUID().toString();
        doc.list = Arrays.asList("bbb", "aaa");
        mongoTemplate.insert(doc, "testDocument");

        // searching for document using $all
        TestDocument found1 = mongoTemplate.findOne(new BasicQuery("{ list: {$all: [ \"aaa\", \"bbb\" ]} }"), TestDocument.class);
        Assertions.assertThat(found1).isNotNull();
        Assertions.assertThat(found1.id).isEqualTo(doc.id);

        // searching for document using $size
        TestDocument found2 = mongoTemplate.findOne(new BasicQuery("{ list: { $size: 2} }"), TestDocument.class);
        Assertions.assertThat(found2).isNotNull();
        Assertions.assertThat(found2.id).isEqualTo(doc.id);

        // searching for document using $and
        TestDocument found3 = mongoTemplate.findOne(new BasicQuery("{ $and: [{ list: { $size: 2} }, { list: {$all: [ \"aaa\", \"bbb\" ]} }] }"), TestDocument.class);
        Assertions.assertThat(found3).isNotNull();
        Assertions.assertThat(found3.id).isEqualTo(doc.id);

        // searching for document using $size + $all
        TestDocument notFound = mongoTemplate.findOne(new BasicQuery("{ list: {$all: [ \"aaa\", \"bbb\" ], $size: 2} }"), TestDocument.class);
        // will always fails
        Assertions.assertThat(notFound).isNotNull();
    }

    @Document
    public static class TestDocument {

        @Id
        String id;
        List<String> list;

    }

}

```
Best regards
Piotr

**Comment by bwaldvogel** [comment]
Thanks for reporting this bug. It should be fixed with 6d18f7bd16ea2ca19183c4320ded690d7c784244.
Could you re-test with your other queries on this version?

If it helps, you can build a ["fat" JAR as described in the README](https://github.com/bwaldvogel/mongo-java-server#building-a-fat-jar-that-contains-all-dependencies).

**Comment by piotrrzacki** [comment]
Thanks for fix, but it is not working entirely.
If $all criteria is first, then works as expected, if $size is first criteria, then it returns document not matching criteria. Unit test:

```java
@Test
    public void contextLoads() {
        // insert document with field as list
        TestDocument doc = new TestDocument();
        doc.id = UUID.randomUUID().toString();
        doc.list = Arrays.asList("bbb", "aaa");
        mongoTemplate.insert(doc, "testDocument");

        // searching for document using $all + $size for aaa (value exists in saved document) and ccc (not exists in saved document) -> query returns null
        TestDocument notFound1 = mongoTemplate.findOne(new BasicQuery("{ list: {$all: [ \"aaa\", \"ccc\" ]}, $size: 2 }"), TestDocument.class);
       // searching for document using $size + $all for aaa (value exists in saved document) and ccc (not exists in saved document) -> query returns saved document - but it shouldn't
        TestDocument notFound2 = mongoTemplate.findOne(new BasicQuery("{ list: {$size: 2, $all: [ \"aaa\", \"ccc\" ]} }"), TestDocument.class);

        Assertions.assertThat(notFound1).isNull();
        // fails at this point
        Assertions.assertThat(notFound2).isNull();
    }
```

**Comment by bwaldvogel** [comment]
It should be fixed with 34aed094525c2c6dacfc482ad3c49e4f1965caf9. Can you re-test again?

**Comment by piotrrzacki** [comment]
After last commit it works as expected. Thank you.

**Comment by bwaldvogel** [comment]
The fix is part of version `1.21.0`.
